tools/build-tool/data/android/integrated-services/clientserver/app/src/main/java/net/sourceforge/castleengine/ServiceClientServer.java

```diff
@@ -0,0 +1,270 @@
+/*
+  Copyright 2018 Benedikt Magnus.
+
+  This file is part of "Castle Game Engine".
+
+  "Castle Game Engine" is free software; see the file COPYING.txt,
+  included in this distribution, for details about the copyright.
+
+  "Castle Game Engine" is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+
+  ----------------------------------------------------------------------------
+*/
+
+package net.sourceforge.castleengine;
+
+import android.os.Handler;
+import android.os.Looper;
+import android.util.Log;
+import android.content.Context;
+
+import java.net.*;
+import java.io.*;
+import java.util.*;
+import java.util.concurrent.*;
+
+public class ServiceClientServer extends ServiceAbstract
+{
+    private static final String CATEGORY = "ServiceClientServer";
+
+    private class ServerTuple
+    {
+        public Boolean active = true;
+        public ConcurrentHashMap<String, Boolean> activeMap = null;
+
+        public ServerTuple(ConcurrentHashMap<String, Boolean> setActiveMap)
+        {
+            activeMap = setActiveMap;
+        }
+    }
+
+    private ConcurrentHashMap<String, ConcurrentHashMap<String, List<String>>> messageMap = new ConcurrentHashMap<String, ConcurrentHashMap<String, List<String>>>();
+    private ConcurrentHashMap<String, ServerTuple> activeMap = new ConcurrentHashMap<String, ServerTuple>();
+
+    public ServiceClientServer(MainActivity activity)
+    {
+        super(activity);
+    }
+
+    public String getName()
+    {
+        return "serviceclientserver";
+    }
+
+    @Override
+    public boolean messageReceived(String[] parts)
+    {
+        if ((parts.length < 4) || (parts.length > 6) || !parts[0].equals("serviceclientserver")) //serviceclientserver key action param1 (param2) (param3)
+            return false;
```

Comment:
The message string should be `client-server`, so

```
...parts[0].equals("client-server")...
```

With hyphens, and without initial `service` --- for consistency with other messages, we use hyphens there for multi-word strings. All the code sending/receiving messages should be adjusted to it too, naturally.